src/day8/re_frame/trace.cljs

```diff
@@ -215,43 +216,61 @@
                                           tag-str))]]
                                [:td.trace--meta.trace--details-icon
                                   {:on-click #(.log js/console tags)}]]))))))))
+
 (defn render-trace-panel []
   (let [filter-input               (r/atom "")
         filter-items               (r/atom (localstorage/get "filter-items" []))
         filter-type                (r/atom :contains)
         input-error                (r/atom false)
+        categories                 (r/atom #{:event :sub/run :sub/create})
         trace-detail-expansions    (r/atom {:show-all? false :overrides {}})]
     (add-watch filter-items
                :update-localstorage
                (fn [_ _ _ new-state]
                  (localstorage/save! "filter-items" new-state)))
     (fn []
-      (let [visible-traces       (if (= @filter-items [])
-                                   @traces
-                                   (filter (apply every-pred (map query->fn @filter-items)) @traces))
+      (let [toggle-category-fn   (fn [category-keys]
+                                   (swap! categories #(set (if (empty? (set/intersection % category-keys))
+                                                             (concat category-keys %)
+                                                             (remove category-keys %)))))
```

Comment:
You could use `clojure.set/union` and `clojure.set/difference` here. 
- It would make it more apparent that you're working with sets
- you wouldn't need the wrapping `set` call
- they're faster as far as I remember (not really important)

Also just to recap what the function does (and to make sure I understand it correctly 🙂) :
- **if** the newly toggled `category-keys` are **not** already in the `categories` atom
- **then** just merge the two sets
- **else** remove the newly toggled `category-keys`

With that in mind you could consider using the `set/subset?` or `set/superset?` predicates which would save you the `empty?` call.

Reply:
that makes a lot of sense, thanks for your feedback :) I changed the code to use set operations instead!